rust/opendp/src/trans/mod.rs

```diff
@@ -0,0 +1,246 @@
+}
+
+fn clamp<T: Copy + PartialOrd>(lower: T, upper: T, x: &Vec<T>) -> Vec<T> {
+    fn clamp1<T: Copy + PartialOrd>(lower: T, upper: T, x: T) -> T {
+        if x < lower { lower } else if x > upper { upper } else { x }
+    }
+    x.into_iter().map(|e| clamp1(lower, upper, *e)).collect()
+}
+
+pub struct BoundedSum<MI, T> {
+    input_metric: PhantomData<MI>,
+    data: PhantomData<T>,
+}
+
+impl<MO, T> MakeTransformation3<VectorDomain<IntervalDomain<T>>, AllDomain<T>, HammingDistance, MO, T, T, MO> for BoundedSum<HammingDistance, T>
+    where T: 'static + Copy + PartialOrd + Sub<Output=T> + NumCast + Mul<Output=T> + Sum<T>,
+          MO: SensitivityMetric<Distance=T> {
+    fn construct(lower: T, upper: T, output_metric: MO) -> Transformation<VectorDomain<IntervalDomain<T>>, AllDomain<T>, HammingDistance, MO> {
+        Transformation::new(
+            VectorDomain::new(IntervalDomain::new(Bound::Included(lower.clone()), Bound::Included(upper.clone()))),
+            AllDomain::new(),
+            |arg: &Vec<T>| arg.iter().cloned().sum(),
+            HammingDistance::new(),
+            output_metric,
+            move |d_in: &u32, d_out: &T| *d_out >= T::from(*d_in).unwrap() * (upper - lower))
+    }
+}
+
+// TODO: this is kind of ugly and should bubble results
+fn max<T: PartialOrd>(a: T, b: T) -> T {
+    match a.partial_cmp(&b) {
+        Some(Ordering::Less) => b,
+        _ => a
+    }
+}
+
+impl<MO, T> MakeTransformation3<VectorDomain<IntervalDomain<T>>, AllDomain<T>, SymmetricDistance, MO, T, T, MO> for BoundedSum<SymmetricDistance, T>
+    where T: 'static + Copy + PartialOrd + Sub<Output=T> + NumCast + Mul<Output=T> + Sum<T> + Signed,
+          MO: SensitivityMetric<Distance=T> {
+    // Question- how to set the associated type for a trait that a concrete type is using
+    fn construct(lower: T, upper: T, output_metric: MO) -> Transformation<VectorDomain<IntervalDomain<T>>, AllDomain<T>, SymmetricDistance, MO> {
+        Transformation::new(
+            VectorDomain::new(IntervalDomain::new(Bound::Included(lower.clone()), Bound::Included(upper.clone()))),
+            AllDomain::new(),
+            |arg: &Vec<T>| arg.iter().cloned().sum(),
+            SymmetricDistance::new(),
+            output_metric,
+            // d_out >= d_in * max(|m|, |M|)
+            move |d_in: &u32, d_out: &T| *d_out >= T::from(*d_in).unwrap() * max(num::abs(lower), num::abs(upper)))
+    }
+}
+
+impl<MO, T> MakeTransformation4<SizedDomain<VectorDomain<IntervalDomain<T>>>, AllDomain<T>, SymmetricDistance, MO, usize, T, T, MO> for BoundedSum<SymmetricDistance, T>
+    where T: 'static + Copy + PartialOrd + Sub<Output=T> + NumCast + Mul<Output=T> + Div<Output=T> + Sum<T>,
+          MO: SensitivityMetric<Distance=T>,
+          SymmetricDistance: Metric<Distance=u32>  {
+    fn construct(length: usize, lower: T, upper: T, output_metric: MO) -> Transformation<SizedDomain<VectorDomain<IntervalDomain<T>>>, AllDomain<T>, SymmetricDistance, MO> {
```

Comment:
It's a bummer that we have to pass in output_metric as a parameter here. Could we not do the same trick as with DatasetMetric, have a new() associated function?

Reply:
I think the reason we didn't add `new()` to SensitivityMetric was because some sensitivities take parameters.

Reply:
The doubt here was that the new() constructor would not have sufficient information to create:
`struct LpSensitivity {p: u32}`
I agree that there is a downside for not defining new for all metrics, because you get inconsistent interfaces-- sometimes you need to pass metrics, sometimes you don't.

Reply:
Ah yes, I forgot about that source of variation.

Is this something that's actually used by some mechanisms? Or in practice, would we only have say L1, L2, L-infinity norm?

Reply:
It's more a question of how general we want to be. I think the vast majority of cases will need L1 or L2 norms. In cases where we need Lp norms, the default constructor can just let p = 1. Unintentionally setting p = 1 would still get caught by runtime checks when chaining. Constructors that emit or accept sensitivity in Lp space can take an extra LpSensitivity argument, or p argument. 

Interesting note, when we start having non-empty metric structs, we can no longer omit the check for metric equality when chaining.

I pushed a commit that adds a `new()` constructor to the `SensitivityMetric` trait.

Reply:
Makes sense, thanks!